Add experimental metric attributes advice API (#3546)
open-telemetry/opentelemetry-specification: specification/metrics/api.md

    * `ExplicitBucketBoundaries` (`double[]`): The recommended set of bucket
      boundaries to use if aggregating to
      [explicit bucket Histogram metric data point](./data-model.md#histogram).
  * All instruments:

>     * `Attributes` (a list of [attribute keys](../common/README.md#attribute)):
>       The recommended set of attribute keys to keep when recording a metric
>       measurement. Attribute values whose corresponding keys do not appear in
>       this list will be discarded.

Reviewer: The existing views attribute parameters has [issues](https://github.com/open-telemetry/opentelemetry-specification/issues/3441) with its limited functionality. Instead of copying the existing views parameter here, can we define this as an attribute filtering function? Maybe something like this:

```suggestion
    * `Attributes` (an attribute filtering function that takes an
      [attribute](../common/README.md#attribute) and returns a
      filtering decision): The attribute filtering function used when
      recording a metric measurement. When the function returns
      a decision to drop for an attribute key-value it is passed, that
      attribute will be discarded.
```

Reviewer 2: This is a different capability that what is described here and can get the user in trouble. Consider an instrument recording values for three series: {key: a}, {key: b}, and {key: c}

A user could define a predicate like:
```
....
.setAdvice(advice ->
  advice.setAttributes((key, value) => {
    if (key.equals("a") || key.equals("b")) {
      return true;
    }
    return false;
  });
```
This would result in the following series: {key: a}, {key: b}, {}

Within a single instrument, its normal that all series have the same set of keys. Making it easy to have keys which vary from series to series is a foot gun.

What are the use cases we aim to support through this increased flexibility?

Reviewer: I'm not sure I follow the reasoning this is a foot gun. The filter only applies to a single instrument, right?

As for things the current approach does not address, you only have an allow list of keys. It does not support:

- A deny-list of keys (#3441)
- A deny-list of values (i.e. sentinel values)
- An allow-list of values (i.e. paths without sensitive information in them)
- Higher order matching functions like regular expressions or other complex string matching for keys
- Ability to filter attributes with unsupported value types (i.e. a backend does not support array values)

There are almost certainly more use cases, but that's kinda the point. Allowing an open ended user function gives users the ability to program the functionality they need.

Reviewer: > A user could define a predicate like:
> 
> ```
> ....
> .setAdvice(advice ->
>   advice.setAttributes((key, value) => {
>     if (key.equals("a") || key.equals("b")) {
>       return true;
>     }
>     return false;
>   });
> ```
> 
> This would result in the following series: {key: a}, {key: b}, {}

Isn't this just the equivalent of:

 ```
....
.setAdvice(advice ->
  advice.setAttributes(filterOnTheseKeys={"a","b"});
```

I feel like I'm missing something here...

Reviewer 2: Oops typo in the example (updated). In this example the attribute's key is "key" - a, b, and c are values. Sorry about the confusing example 😁.

So with preserve attributes you could do:
```
....
.setAdvice(advice ->
  advice.preserveAttributes(List.of("key"))
  // or if predicates are allowed
  advice.preserveAttributes(key -> key.equals("key"))
```

But you can't have the behavior change based on the value.

Reviewer 2: > A deny-list of keys (https://github.com/open-telemetry/opentelemetry-specification/issues/3441)
A deny-list of values (i.e. sentinel values)
An allow-list of values (i.e. paths without sensitive information in them)
Higher order matching functions like regular expressions or other complex string matching for keys
Ability to filter attributes with unsupported value types (i.e. a backend does not support array values)

I like the examples that give you more flexibility restricting keys (deny list of keys, regular expressions on keys), and a predicate which accepts a key would solve these.

But not sure the examples related to values make sense to me in advice. Advice is provided at the point of instrumentation. If you know something about which values should be denied or accepted, or that certain value types are unsupported, shouldn't you just incorporate that into the measurements you record? I suppose advice is meant to have the potential to be overridden by sdk configuration like views. So the way to think about these value examples would be something like: "advice adds a default deny list of values, but the user can override this behavior in the SDK if needed". That presumes you have an SDK mechanism to override the advice applying to attribute values, which doesn't exist today, but is related to #3550 and #3545.

Reviewer: > So the way to think about these value examples would be something like: "advice adds a default deny list of values, but the user can override this behavior in the SDK if needed". That presumes you have an SDK mechanism to override the advice applying to attribute values, which doesn't exist today, but is related to #3550 and #3545.

Yeah, that sounds correct to me.

Reviewer: I think the example @lmolkova provided about the limiting of HTTP method values is a good motivator for including the value in the filter function. The advice could state to only use known values, but if a user wants to override some HTTP instrumentation library because they use non-standard methods values they could do so with a view (depending, like you said, on #3550 or something similar). This probably also applies to non-standard HTTP status codes.

Author: > * A deny-list of keys ([Support dropping of specific attributes from instruments #3441](https://github.com/open-telemetry/opentelemetry-specification/issues/3441))
> * Higher order matching functions like regular expressions or other complex string matching for keys

I'd argue that the instrumentation usually knows the attribute keys it recommends to use as metric attributes - it has to choose low cardinality attributes for that purpose after all.
The #3441 issue is a use case for views, not the advice API.

> * Ability to filter attributes with unsupported value types (i.e. a backend does not support array values)

Instrumentation *must not* be hard-coded against a particular backend. This is a view API use case.

> * A deny-list of values (i.e. sentinel values)
> * An allow-list of values (i.e. paths without sensitive information in them)

I'm not sure whether we actually have any real use case for these; the HTTP method issue is off the table here, since we're not gonna use advice API to solve it.

After some thought, I think I'm against making this a function. The advice API is not supposed to be a general data manipulation mechanism for end users, it's supposed to relay the recommendations of instrumentation authors to the SDK. The end users have the views API.

Reviewer: Is [this](https://github.com/open-telemetry/opentelemetry-specification/pull/3546#issuecomment-1582748626) comment's motivation for this PR still valid then?

Author: Yes, it is.
The advice API will be used by the instrumentation authors to limit the attributes to whatever semantic conventions require; and if the end user wishes to add/remove/modify the attributes they'll be able to do that using the views API.

I think I should've made a better explanation of the advice/views (API/SDK) symmetry in the PR description; my point is that the advice API is something that is intended to be used primarily by the instrumentation authors. The end user will have a possibility to override the instrumentation authors' decision via views (this is not the case today, i.e. the end user cannot easily add more attributes to metrics).

Reviewer 3: ```suggestion
    * `Attributes` (a list of [attribute keys](../common/README.md#attribute)):
      The recommended set of attribute keys to be used for the resulting metrics.
```
